### ISSUE TITLE
Reduce the number of blocked threads in the application

### ISSUE THREAD

**Issue by JaidenAshmore** [JaidenAshmore/java-dynamic-sqs-listener#106]
Currently the core application will have threads blocked while they wait for messages and therefore if there are 10 listeners with 10 levels of concurrency there are 100 threads sitting their idle until messaages are there to process. It would be great if the threads are only running when requesting messages and processing them not when they are waiting.

This can be completed by using a more asynchronous approach to the application.

**Comment by JaidenAshmore** [comment]
I experimented on this and tested whether there were any performance gains by converting to a more asynchronous model. The test included placing 500 thousands messages onto the ElasticMQ and seeing how long it took to process them all with each approach.  The performance difference between the two approaches were negligible (if anything asynchronous was slightly slower).  Converting to this asynchronous model also makes the code more complicated and harder to read and having threads blocked while it waits for messages is not the biggest deal.

The changes that I made were off of the current commit which is a1bdde0d3c96d71997c3001cecc3ddba0d98c709 and are included in the following patch: 
[106_message_retriever_return_completable_futures.zip](https://github.com/JaidenAshmore/java-dynamic-sqs-listener/files/3245324/106_message_retriever_return_completable_futures.zip)

Note that my implementation was still quite naive and was just a spike so maybe I was implementing it incorrectly so maybe it is able to be improved to be more performant. I potentially will take a look at this again in the future.
